File: 001.BOGGLE/BOGGLE.py
#!/usr/bin/env python
gnWidth = 7
gnHeight = 7
lstCoord = [ -gnWidth-1, -gnWidth-0, -gnWidth+1, -1, +1, gnWidth-1, gnWidth+0, gnWidth+1]

def genSquare(strBoard, nIdx):
    i = 0
    while(i < 8):
        nCoord = lstCoord[i]
        nTarget = nIdx + nCoord
        yield nTarget, strBoard[nTarget]
        i += 1
    return

def step(strBoard, nStart, strWord, nCur):
    if nCur == len(strWord):
        return True

    idx = strBoard.find(strWord[nCur])
    while (idx > 0):
        if ( idx != nStart ):
            for i in lstCoord :
                if ( idx == (nStart+i) ):
                   if step(strBoard, idx, strWord, nCur+1) == True:
                       return True
                   else:
                        break
        idx = strBoard.find(strWord[nCur], idx+1)

    return False

def doBoggleGame(strBoard, lstWords):
    for strWord in lstWords:
        fGO = True
        for C in strWord:
            if strBoard.find(C) < 0:
                print "%s NO" % (strWord)
                fGO = False
                break

        if fGO :
            idx = strBoard.find(strWord[0])
            while (idx > 0):
                if step(strBoard, idx, strWord, 1) == True:
                    print "%s YES" % (strWord)
                    break
                idx = strBoard.find(strWord[0], idx+1)

            if idx < 0 :
                print "%s NO" % (strWord)

def main():
    strBoard = ""
    nStageCount = raw_input()
    for i in xrange(0, int(nStageCount)) :
        strBoard += '0' * gnWidth
        for n in xrange(0, 5) :
            strBoard += '0'
            strBoard += raw_input()
            strBoard += '0'
        strBoard += '0' * gnWidth

        lstWords = []
        nWordCount = raw_input()
        for n in xrange(0, int(nWordCount)) :
            lstWords.append(raw_input())

        doBoggleGame(strBoard, lstWords)

if __name__ == "__main__" :
    main()
    

File: 001.BOGGLE/BOGGLE.cpp
#include <iostream>
#include <cstring>

const int gnWidth = 7;
const int gnHeight = 7;
const int gnSize = gnWidth*gnHeight;
const int garrCoord[8] = 
{
    -gnWidth - 1,
    -gnWidth - 0,
    -gnWidth + 1,
             - 1,
             + 1,
    +gnWidth - 1,
    +gnWidth + 0,
    +gnWidth + 1
};

char* find(char *pBegin, char *pEnd, char nVal)
{
    while (pBegin != pEnd)
    {
        if (*pBegin == nVal) return pBegin;
        ++pBegin;
    }

    return pEnd;
}

bool step(char *pBoard, char *pStart, char *szWord, char nLength, int nCur)
{
    if (nLength == nCur)
        return true;
    
    for (int i = 0; i < 8; i++)
    {
        char *pPos = pStart + garrCoord[i];
        if (*pPos == szWord[nCur])
            if (step(pBoard, pPos, szWord, nLength, nCur+1))
                return true;
    }
    return false;
}

void doBoggleGame(char *pBoard, char arrWord[10][11], int nWordCount)
{
    char *pBegin = pBoard;
    char *pEnd = pBoard+gnSize;

    for (int nWC = 0; nWC < nWordCount; nWC++)
    {
        std::cout << arrWord[nWC] << std::endl;
        int nLength = strlen(arrWord[nWC]);
        while (--nLength)
        {
            if (find(pBegin, pEnd, arrWord[nWC][nLength]) == pEnd)
            {
                std::cout << arrWord[nWC] << " NO" << std::endl;
            }
        }

        nLength = strlen(arrWord[nWC]);
        char *pStart = find(pBegin, pEnd, arrWord[nWC][0]);
        while (pStart != pEnd)
        {
            if (step(pBoard, pStart, arrWord[nWC], nLength, 1))
            {
                std::cout << arrWord[nWC] << " YES" << std::endl;
                break;
            }
            pStart = find(pStart+1, pEnd, arrWord[nWC][0]);
        }

        if (pStart == pEnd)
            std::cout << arrWord[nWC] << " NO" << std::endl;
    }
}

int main (int argc, char** argv)
{
    int nStageCount = 0;
    int nWordCount = 0;
    char arrBoard[gnSize];
    char arrWord[10][11];

    std::cin >> nStageCount;
    while (nStageCount--)
    {
        memset (arrBoard, 0, sizeof(char)*gnSize);
        char *pStage = arrBoard + gnWidth + 1;
        for (int i = 0; i < 5; i++)
        {
            std::cin >> (pStage + i*gnWidth);
        }

        std::cin >> nWordCount;
        for (int i = 0; i < nWordCount; i++)
        {
            std::cin >> arrWord[i];
        }

        doBoggleGame(arrBoard, arrWord, nWordCount);
    };

    return 0;
}
